// @vuepress/bundler-vite/lib/build/build.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.build = void 0;
const utils_1 = require("@vuepress/utils");
const vite_1 = require("vite");
const resolveViteConfig_1 = require("../resolveViteConfig");
const renderPage_1 = require("./renderPage");
const build = async (options, app) => {
    // plugin hook: extendsBundlerOptions
    await app.pluginApi.hooks.extendsBundlerOptions.process(options, app);
    // vite compile
    let clientOutput;
    let serverOutput;
    await (0, utils_1.withSpinner)('Compiling with vite')(async () => {
        // create vite config
        const clientConfig = await (0, resolveViteConfig_1.resolveViteConfig)({
            app,
            options,
            isBuild: true,
            isServer: false,
        });
        const serverConfig = await (0, resolveViteConfig_1.resolveViteConfig)({
            app,
            options,
            isBuild: true,
            isServer: true,
        });
        [clientOutput, serverOutput] = await Promise.all([
            (0, vite_1.build)(clientConfig),
            (0, vite_1.build)(serverConfig),
        ]);
    });
    // render pages
    await (0, utils_1.withSpinner)('Rendering pages')(async () => {
        // load ssr template file
        const ssrTemplate = (await utils_1.fs.readFile(app.options.templateBuild)).toString();
        // get client bundle entry chunk and css asset
        const clientEntryChunk = clientOutput.output.find((item) => item.type === 'chunk' && item.isEntry);
        const clientCssAsset = clientOutput.output.find((item) => item.type === 'asset' && item.fileName.endsWith('.css'));
        // get server bundle entry chunk
        const serverEntryChunk = serverOutput.output.find((item) => item.type === 'chunk' && item.isEntry);
        // load the compiled server bundle
        const serverEntryPath = app.dir.dest('.server', serverEntryChunk.fileName);
        // delete server entry cache to allow building multiple times
        // in the same dest dir
        delete require.cache[serverEntryPath];
        const { createVueApp } = require(serverEntryPath);
        // create vue ssr app
        const { app: vueApp, router: vueRouter } = await createVueApp();
        // pre-render pages to html files
        const spinner = (0, utils_1.ora)();
        for (const page of app.pages) {
            spinner.start(`Rendering pages ${utils_1.chalk.magenta(page.path)}`);
            await (0, renderPage_1.renderPage)({
                app,
                page,
                vueApp,
                vueRouter,
                ssrTemplate,
                output: clientOutput.output,
                outputEntryChunk: clientEntryChunk,
                outputCssAsset: clientCssAsset,
            });
        }
        spinner.stop();
    });
    // keep the server bundle files in debug mode
    if (!app.env.isDebug) {
        // remove server dest directory after pages rendered
        await utils_1.fs.remove(app.dir.dest('.server'));
    }
};
exports.build = build;
